readthedocs/projects/views/public.py

```diff
@@ -190,10 +192,18 @@ def project_download_media(request, project_slug, type_, version_slug):
     )
     privacy_level = getattr(settings, 'DEFAULT_PRIVACY_LEVEL', 'public')
     if privacy_level == 'public' or settings.DEBUG:
-        path = os.path.join(
-            settings.MEDIA_URL, type_, project_slug, version_slug,
-            '%s.%s' % (project_slug, type_.replace('htmlzip', 'zip')))
-        return HttpResponseRedirect(path)
+        storage_path = version.project.get_storage_path(type_=type_, version_slug=version_slug)
+        if storage.exists(storage_path):
```

Comment:
In the case where the item is not in storage but is on disk, this will result in an extra network call. `storage.exists` results in an API call to Azure (or wherever storage is backed). However, it should be reasonably fast since it's in the same data center. In the case where media is backed by the local file system, this is just an `os.path.isfile` so not a big deal.

Reply:
That seems fine, and hopefully we will get files into the storage quickly. I'd be 👍 on syncing all the media files after we ship the code, so that we rarely hit this case, as well.

Reply:
I did run into an issue locally, where I have a project with ~100 active versions. It required 100 calls to Azure in order to load the project downloads page. This feels less than ideal -- I think the only way to get around it is to denormalize the "has_pdf" state onto the Version model.

Reply:
That seems suboptimal. Perhaps we could check whether the last build for a version built PDFs and was successful?

Reply:
Would it also make sense to cache this lookup then?

Reply:
> Would it also make sense to cache this lookup then?

My guess is no. Checking whether any specific epub or pdf exists is not very common.